src/app/base/components/LimitedIpInput/LimitedIpInput.tsx

```diff
@@ -0,0 +1,99 @@
+import { useEffect, useState } from "react";
+
+import { Input } from "@canonical/react-components";
+import type { InputProps } from "@canonical/react-components";
+import { useFormikContext } from "formik";
+
+import type { Subnet } from "@/app/store/subnet/types";
+import {
+  getImmutableAndEditableOctets,
+  getIpRangeFromCidr,
+} from "@/app/utils/subnetIpRange";
+
+type Props = Omit<InputProps, "maxLength" | "placeholder" | "type" | "name"> & {
+  cidr: Subnet["cidr"];
+  name: string;
+};
+
+const LimitedIpInput = ({ cidr, name, ...props }: Props) => {
+  const [startIp, endIp] = getIpRangeFromCidr(cidr);
+  const [immutable, editable] = getImmutableAndEditableOctets(startIp, endIp);
+  const [immutableWidth, setImmutableWidth] = useState("inherit");
+  const [inputWrapper, setInputWrapper] = useState<Element | null>(null);
+  const [inputElement, setInputElement] = useState<Element | null>(null);
+
+  const formikProps = useFormikContext();
+
+  const getMaxLength = () => {
+    const immutableOctetsLength = immutable.split(".").length;
+
+    if (immutableOctetsLength === 3) {
+      return 3; // 3 digits, no dots
+    } else if (immutableOctetsLength === 2) {
+      return 7; // 6 digits, 1 dot
+    } else if (immutableOctetsLength === 1) {
+      return 11; // 9 digits, 2 dots
+    } else {
+      return 15; // 12 digits, 3 dots
+    }
+  };
+
+  useEffect(() => {
+    setInputWrapper(document.querySelector(".limited-ip-input__input-wrapper"));
+    setInputElement(document.querySelector(".limited-ip-input__input"));
```

Comment:
The only problem with using `ref` is that the element we're trying to access is the wrapper of the input, rather than the input itself. `react-components` doesn't let us pass a `ref` through to the wrapper, only a classname

Reply:
Nvm, figured it out - we can put the ref on the top-level div that wraps everything, and then access children from there.